File: cci/01_03_urlify.py
# -*- coding: utf-8 -*-
# @Author: krocki
# @Date:   2017-01-14 19:15:43
# @Last Modified by:   krocki
# @Last Modified time: 2017-01-15 15:06:58

#urlify - replace space with %20
#TODO: write in C

# def nowhitespace(k):

# 	if k == ' ': 
# 		return '%20'
# 	else: 
# 		return k

# def urlify(str):

# 	list_str = list(str);
# 	processed = [nowhitespace(k) for k in list_str]
# 	# return reduce(lambda x,y: x + y, processed)
# 	return ''.join(processed)

def count_spaces(array, l):

	counter = 0
	
	for i, a in enumerate(array):
		if i >= l: break
		if a == ' ': counter += 1

	return counter

def urlify (array, l):

	num_spaces = count_spaces(array, l)

	new_length = (l - num_spaces) + 3 * num_spaces

	k = l

	for i in range(new_length-1, 0, -1):
		array[i] = array[k]
		print array[k]
		k = k - 1
		if (array[k] == ' '):
			array[i-2] = '%'
			array[i-1] = '2'
			array[i] = '0'
			i = i - 2

	return array

print urlify("abc ads asdddd      ", 14)

File: cci/01_03_urlify.c
/*
* @Author: krocki
* @Date:   2017-01-15 15:09:06
* @Last Modified by:   krocki
* @Last Modified time: 2017-01-15 15:24:50
*/

#include <stdio.h>

int count_spaces(char* arr, int len) {

	int i;
	int counter = 0;

	for (i = 0; i < len; i++)
		if (arr[i] == ' ') counter++;

	return counter;

}

void urlify(char* arr, int len) {

	int num_spaces = count_spaces(arr, len);
	int extended_length = len + num_spaces * 2;

	int i = extended_length - 1;
	int k;

	for (k = len - 1; k >= 0; k--) {

		if (arr[k] != ' ') {

			arr[i] = arr[k];
			--i;

		} else {

			arr[i] = '0';
			arr[i - 1] = '2';
			arr[i - 2] = '%';
			i -= 3;

		}

	}

}

int main() {

	char array[] = "abc string zyx    ";
	int len = 14;

	urlify(array, len);

	int i;

	printf("%s\n", array);
	return 0;
}
